pageserver/src/layered_repository.rs

```diff
@@ -1754,135 +1864,47 @@ impl LayeredTimeline {
             // might be referenced by child branches forever.
             // We can track this in child timeline GC and delete parent layers when
             // they are no longer needed. This might be complicated with long inheritance chains.
-            for retain_lsn in &retain_lsns {
+            for retain_lsn in retain_lsns {
                 // start_lsn is inclusive
-                if &l.get_start_lsn() <= retain_lsn {
+                if &l.get_lsn_range().start <= retain_lsn {
                     info!(
-                        "keeping {} {}-{} because it's still might be referenced by child branch forked at {} is_dropped: {} is_incremental: {}",
-                        seg,
-                        l.get_start_lsn(),
-                        l.get_end_lsn(),
+                        "keeping {} because it's still might be referenced by child branch forked at {} is_dropped: xx is_incremental: {}",
+                        l.filename().display(),
                         retain_lsn,
-                        l.is_dropped(),
                         l.is_incremental(),
                     );
-                    if seg.rel.is_relation() {
-                        result.ondisk_relfiles_needed_by_branches += 1;
-                    } else {
-                        result.ondisk_nonrelfiles_needed_by_branches += 1;
-                    }
+                    result.layers_needed_by_branches += 1;
                     continue 'outer;
                 }
             }
 
             // 3. Is there a later on-disk layer for this relation?
-            if !l.is_dropped()
-                && !layers.newer_image_layer_exists(
-                    l.get_seg_tag(),
-                    l.get_end_lsn(),
-                    disk_consistent_lsn,
-                )
-            {
+            //
+            // The end-LSN is exclusive, while disk_consistent_lsn is
+            // inclusive. For example, if disk_consistent_lsn is 100, it is
+            // OK for a delta layer to have end LSN 101, but if the end LSN
+            // is 102, then it might not have been fully flushed to disk
+            // before crash.
+            if !layers.newer_image_layer_exists(
+                &l.get_key_range(),
+                l.get_lsn_range().end,
+                disk_consistent_lsn + 1,
+            )? {
```

Comment:
We should make sure that all deltas we want to retain are **preceded** by an image that we keep. This check is not enough. For example, consider this sequence of layers for a given fixed key range:
1. image
2. delta
3. image

If the cutoff is at 2, we can't delete 1 even though there's a newer image. If we remove it then the delta becomes useless.

Reply:
Good point. This sounded familiar, and I realized that we've seen this before: https://github.com/zenithdb/zenith/issues/707. So this isn't new, but yeah we really should fix this...

Reply:
ok. Maybe just link to the issue, no need to fix. Save the readers some time, it takes a while to triple check if this is actually broken, and then look for existing issues.